narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,381 @@
+        let name = self.name.clone();
+        let network = self.network.clone();
+        let committee = self.committee.clone();
+        let certificate_store = self.certificate_store.clone();
+        let tx_certificates_loopback = self.tx_certificates_loopback.clone();
+        let metrics = self.metrics.clone();
+
+        tokio::task::spawn(async move {
+            // Send request to fetch certificates.
+            let request = FetchCertificatesRequest {
+                progression: progression.into_iter().collect(),
+                max_items: MAX_CERTIFICATES_TO_FETCH,
+            };
+            let response = fetch_certificates_helper(
+                name.clone(),
+                network.clone(),
+                committee.clone(),
+                request.clone(),
+            )
+            .await;
+
+            // Process and store fetched certificates.
+            store_certificates_helper(
+                response,
+                committee,
+                &certificate_store,
+                tx_certificates_loopback,
+                metrics,
+            )
+            .await?;
+
+            Ok(())
+        })
+    }
+
+    #[allow(clippy::mutable_key_type)]
+    fn read_round_progression(&self) -> DagResult<BTreeMap<PublicKey, Round>> {
+        let mut progression = BTreeMap::new();
+        for (name, _) in self.committee.authorities() {
+            // Last round is 0 (genesis) when authority is not found in store.
+            let last_round = self.certificate_store.last_round_number(name)?.unwrap_or(0);
+            progression.insert(name.clone(), last_round);
+        }
+        Ok(progression)
+    }
+}
+
+/// Fetches certificates from other primaries concurrently, with ~5 sec interval between each request.
+/// Terminates after the 1st successful response is received.
+#[instrument(level = "debug", skip_all)]
+async fn fetch_certificates_helper(
+    name: PublicKey,
+    network: Arc<dyn PrimaryToPrimaryRpc>,
+    committee: Committee,
+    request: FetchCertificatesRequest,
+) -> FetchCertificatesResponse {
+    trace!("Start sending fetch certificates requests");
+    let request_interval = Duration::from_secs(5);
+    let mut peers: Vec<NetworkPublicKey> = committee
+        .others_primaries(&name)
+        .into_iter()
+        .map(|(_, _, network_key)| network_key)
+        .collect();
+    loop {
+        peers.shuffle(&mut ThreadRng::default());
+        let mut fut = FuturesUnordered::new();
+        for peer in peers.iter() {
+            fut.push(network.fetch_certificates(peer, request.clone()));
+            let mut interval = Box::pin(time::sleep(
+                request_interval + Duration::from_millis(ThreadRng::default().gen_range(0..1000)),
+            ));
+            tokio::select! {
+                res = fut.next() => match res {
+                    Some(Ok(resp)) => {
+                        return resp;
+                    }
+                    Some(Err(e)) => {
```

Comment:
I think this will immediately start another pending request if any currently pending request fails, without waiting for the timeout, even if others are still inflight and haven't failed yet. Is that what you want?

Reply:
I think this behavior is desirable in most cases I can think of. In the common case when another node is down, there is no point in waiting to issue another request. Added a comment to make this clearer.